File: server/src/services/rideService.js
const prisma = require('../lib/prisma');
const { getDistance } = require('../utils/geo');
const socketLib = require('../lib/socket');
const logger = require('../lib/logger');
const pricingService = require('./pricingService');
const receiptService = require('./receiptService');

async function createRide({ riderId, pickupLat, pickupLng, destLat, destLng, tier = 'economy', paymentMethod = 'card' }) {
    try {
        const distance = getDistance(pickupLat, pickupLng, destLat, destLng);

        // Calculate surge factor based on demand
        const surgeFactor = await pricingService.calculateSurgeFactor(pickupLat, pickupLng, tier);

        // Calculate fare with surge pricing
        const fareBreakdown = pricingService.calculateFare(distance, tier, surgeFactor);

        const ride = await prisma.ride.create({
            data: {
                riderId,
                pickupLat,
                pickupLng,
                destLat,
                destLng,
                tier,
                paymentMethod,
                distance,
                price: fareBreakdown.totalFare,
                surgeFactor,
                status: 'requested'
            }
        });

        // Update demand metrics
        await pricingService.updateDemandMetrics('increment', 'ride');

        // Notify all online drivers
        const io = socketLib.getIO();
        if (io) io.emit('ride_requested', ride);

        logger.info('Ride created', { rideId: ride.id, tier, surgeFactor, price: ride.price });
        return ride;
    } catch (error) {
        logger.error('Error creating ride', { error: error.message });
        throw error;
    }
}

async function getRideById(id) {
    return prisma.ride.findUnique({
        where: { id },
        include: {
            driver: true,
            receipt: true
        }
    });
}

async function acceptRide(rideId, driverId) {
    try {
        // Transaction to ensure atomicity
        const result = await prisma.$transaction(async (prisma) => {
            const ride = await prisma.ride.findUnique({ where: { id: rideId } });
            if (!ride || ride.status !== 'requested') {
                throw new Error('Ride not available');
            }

            const updateDriver = await prisma.driver.upsert({
                where: { id: driverId },
                update: { status: 'busy' },
                create: {
                    id: driverId,
                    name: `Driver ${driverId}`,
                    status: 'busy'
                }
            });

            const updateRide = await prisma.ride.update({
                where: { id: rideId },
                data: {
                    status: 'assigned',
                    driverId: driverId
                },
                include: { driver: true }
            });

            return { ride: updateRide, driver: updateDriver };
        });

        // Update demand metrics
        await pricingService.updateDemandMetrics('decrement', 'ride');
        await pricingService.updateDemandMetrics('decrement', 'driver');

        const io = socketLib.getIO();
        if (io) io.emit('ride_status_updated', result.ride);

        logger.info('Ride accepted', { rideId, driverId });
        return result.ride;
    } catch (error) {
        logger.error('Error accepting ride', { error: error.message, rideId, driverId });
        throw error;
    }
}

async function startTrip(rideId) {
    try {
        const ride = await prisma.ride.update({
            where: { id: rideId },
            data: {
                status: 'started',
                startedAt: new Date()
            },
            include: { driver: true }
        });

        const io = socketLib.getIO();
        if (io) io.emit('ride_status_updated', ride);

        logger.info('Trip started', { rideId });
        return ride;
    } catch (error) {
        logger.error('Error starting trip', { error: error.message, rideId });
        throw error;
    }
}

async function pauseTrip(rideId) {
    try {
        const ride = await prisma.ride.update({
            where: { id: rideId },
            data: {
                status: 'paused',
                pausedAt: new Date()
            },
            include: { driver: true }
        });

        const io = socketLib.getIO();
        if (io) io.emit('ride_status_updated', ride);

        logger.info('Trip paused', { rideId });
        return ride;
    } catch (error) {
        logger.error('Error pausing trip', { error: error.message, rideId });
        throw error;
    }
}

async function resumeTrip(rideId) {
    try {
        const ride = await prisma.ride.update({
            where: { id: rideId },
            data: {
                status: 'started',
                pausedAt: null
            },
            include: { driver: true }
        });

        const io = socketLib.getIO();
        if (io) io.emit('ride_status_updated', ride);

        logger.info('Trip resumed', { rideId });
        return ride;
    } catch (error) {
        logger.error('Error resuming trip', { error: error.message, rideId });
        throw error;
    }
}

async function endTrip(rideId) {
    try {
        const rideCheck = await prisma.ride.findUnique({ where: { id: rideId } });
        if (!rideCheck) throw new Error('Ride not found');

        const ride = await prisma.ride.update({
            where: { id: rideId },
            data: {
                status: 'completed',
                completedAt: new Date()
            },
            include: { driver: true }
        });

        // Generate receipt
        const fareBreakdown = pricingService.calculateFare(
            ride.distance || 0,
            ride.tier,
            ride.surgeFactor
        );

        await receiptService.generateReceipt(rideId, fareBreakdown);

        // Free up driver
        if (ride.driverId) {
            await prisma.driver.upsert({
                where: { id: ride.driverId },
                update: { status: 'online' },
                create: {
                    id: ride.driverId,
                    name: `Driver ${ride.driverId}`,
                    status: 'online'
                }
            });

            // Update demand metrics
            await pricingService.updateDemandMetrics('increment', 'driver');
        }

        const io = socketLib.getIO();
        if (io) io.emit('ride_status_updated', ride);

        logger.info('Trip ended', { rideId });
        return ride;
    } catch (error) {
        logger.error('Error ending trip', { error: error.message, rideId });
        throw error;
    }
}

async function cancelRide(rideId, reason = 'User cancelled') {
    try {
        const ride = await prisma.ride.update({
            where: { id: rideId },
            data: {
                status: 'cancelled'
            },
            include: { driver: true }
        });

        // Free up driver if assigned
        if (ride.driverId) {
            await prisma.driver.update({
                where: { id: ride.driverId },
                data: { status: 'online' }
            });

            await pricingService.updateDemandMetrics('increment', 'driver');
        }

        // Update demand metrics
        await pricingService.updateDemandMetrics('decrement', 'ride');

        const io = socketLib.getIO();
        if (io) io.emit('ride_status_updated', ride);

        logger.info('Ride cancelled', { rideId, reason });
        return ride;
    } catch (error) {
        logger.error('Error cancelling ride', { error: error.message, rideId });
        throw error;
    }
}

module.exports = {
    createRide,
    getRideById,
    acceptRide,
    startTrip,
    pauseTrip,
    resumeTrip,
    endTrip,
    cancelRide
};

File: server-java/src/main/java/com/ridehailing/service/RideService.java
package com.ridehailing.service;

import com.ridehailing.dto.CreateRideRequest;
import com.ridehailing.dto.FareBreakdown;
import com.ridehailing.model.Driver;
import com.ridehailing.model.Ride;
import com.ridehailing.repository.DriverRepository;
import com.ridehailing.repository.RideRepository;
import com.ridehailing.util.GeoUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.Random;

@Service
@RequiredArgsConstructor
@Slf4j
public class RideService {

    private final RideRepository rideRepository;
    private final DriverRepository driverRepository;
    private final PricingService pricingService;
    private final ReceiptService receiptService;
    private final SocketService socketService;

    /**
     * Create a new ride request
     */
    public Ride createRide(CreateRideRequest request) {
        try {
            // Calculate distance using Haversine formula
            double distance = GeoUtils.getDistance(
                    request.getPickupLat(), request.getPickupLng(),
                    request.getDestLat(), request.getDestLng());

            // Calculate surge factor based on demand
            double surgeFactor = pricingService.calculateSurgeFactor(
                    request.getPickupLat(), request.getPickupLng(),
                    request.getTier() != null ? request.getTier() : "economy");

            // Calculate fare with surge pricing
            String tier = request.getTier() != null ? request.getTier() : "economy";
            FareBreakdown fareBreakdown = pricingService.calculateFare(distance, tier, surgeFactor);

            // Create ride entity
            Ride ride = Ride.builder()
                    .riderId(request.getRiderId())
                    .pickupLat(request.getPickupLat())
                    .pickupLng(request.getPickupLng())
                    .destLat(request.getDestLat())
                    .destLng(request.getDestLng())
                    .tier(tier)
                    .paymentMethod(request.getPaymentMethod() != null ? request.getPaymentMethod() : "card")
                    .distance(distance)
                    .price(fareBreakdown.getTotalFare())
                    .surgeFactor(surgeFactor)
                    .status("requested")
                    .createdAt(LocalDateTime.now())
                    .build();

            ride = rideRepository.save(ride);

            // Update demand metrics
            pricingService.updateDemandMetrics("increment", "ride");

            // Notify all online drivers via WebSocket
            socketService.sendRideRequested(ride);

            log.info("Ride created: rideId={}, tier={}, surgeFactor={}, price={}",
                    ride.getId(), tier, surgeFactor, ride.getPrice());

            return ride;
        } catch (Exception e) {
            log.error("Error creating ride", e);
            throw new RuntimeException("Failed to create ride: " + e.getMessage());
        }
    }

    /**
     * Get ride by ID with related entities
     */
    public Optional<Ride> getRideById(String id) {
        return rideRepository.findById(id);
    }

    /**
     * Driver accepts a ride request
     */
    @Transactional
    public Ride acceptRide(String rideId, String driverId) {
        try {
            // Find and validate ride
            Ride ride = rideRepository.findById(rideId)
                    .orElseThrow(() -> new RuntimeException("Ride not found"));

            if (!"requested".equals(ride.getStatus())) {
                throw new RuntimeException("Ride not available");
            }

            // Find or create driver
            Driver driver = driverRepository.findById(driverId)
                    .orElseGet(() -> Driver.builder()
                            .id(driverId)
                            .name("Driver " + driverId)
                            .status("busy")
                            .createdAt(LocalDateTime.now())
                            .build());

            driver.setStatus("busy");
            driver = driverRepository.save(driver);

            // Generate 4-digit OTP for rider verification
            String otp = String.format("%04d", new Random().nextInt(10000));

            // Update ride with driver assignment and OTP
            ride.setStatus("assigned");
            ride.setDriver(driver);
            ride.setOtp(otp);
            ride = rideRepository.save(ride);

            // Update demand metrics
            pricingService.updateDemandMetrics("decrement", "ride");
            pricingService.updateDemandMetrics("decrement", "driver");

            // Broadcast status update
            socketService.sendRideStatusUpdated(ride);

            log.info("Ride accepted: rideId={}, driverId={}", rideId, driverId);

            return ride;
        } catch (Exception e) {
            log.error("Error accepting ride: rideId={}, driverId={}", rideId, driverId, e);
            throw new RuntimeException("Failed to accept ride: " + e.getMessage());
        }
    }

    /**
     * Start a trip with OTP verification
     */
    public Ride startTrip(String rideId, String otp) {
        try {
            Ride ride = rideRepository.findById(rideId)
                    .orElseThrow(() -> new RuntimeException("Ride not found"));

            // Validate OTP
            if (ride.getOtp() == null || !ride.getOtp().equals(otp)) {
                throw new RuntimeException("Invalid OTP");
            }

            if (!"assigned".equals(ride.getStatus())) {
                throw new RuntimeException("Ride must be assigned before starting");
            }

            ride.setStatus("started");
            ride.setStartedAt(LocalDateTime.now());
            ride.setOtp(null); // Clear OTP after verification
            ride = rideRepository.save(ride);

            socketService.sendRideStatusUpdated(ride);

            log.info("Trip started: rideId={}", rideId);
            return ride;
        } catch (Exception e) {
            log.error("Error starting trip: rideId={}", rideId, e);
            throw new RuntimeException("Failed to start trip: " + e.getMessage());
        }
    }

    /**
     * Pause a trip
     */
    public Ride pauseTrip(String rideId) {
        try {
            Ride ride = rideRepository.findById(rideId)
                    .orElseThrow(() -> new RuntimeException("Ride not found"));

            ride.setStatus("paused");
            ride.setPausedAt(LocalDateTime.now());
            ride = rideRepository.save(ride);

            socketService.sendRideStatusUpdated(ride);

            log.info("Trip paused: rideId={}", rideId);
            return ride;
        } catch (Exception e) {
            log.error("Error pausing trip: rideId={}", rideId, e);
            throw new RuntimeException("Failed to pause trip: " + e.getMessage());
        }
    }

    /**
     * Resume a paused trip
     */
    public Ride resumeTrip(String rideId) {
        try {
            Ride ride = rideRepository.findById(rideId)
                    .orElseThrow(() -> new RuntimeException("Ride not found"));

            ride.setStatus("started");
            ride.setPausedAt(null);
            ride = rideRepository.save(ride);

            socketService.sendRideStatusUpdated(ride);

            log.info("Trip resumed: rideId={}", rideId);
            return ride;
        } catch (Exception e) {
            log.error("Error resuming trip: rideId={}", rideId, e);
            throw new RuntimeException("Failed to resume trip: " + e.getMessage());
        }
    }

    /**
     * End a trip
     */
    @Transactional
    public Ride endTrip(String rideId) {
        try {
            Ride ride = rideRepository.findById(rideId)
                    .orElseThrow(() -> new RuntimeException("Ride not found"));

            ride.setStatus("completed");
            ride.setCompletedAt(LocalDateTime.now());
            ride = rideRepository.save(ride);

            // Generate receipt
            FareBreakdown fareBreakdown = pricingService.calculateFare(
                    ride.getDistance() != null ? ride.getDistance() : 0.0,
                    ride.getTier(),
                    ride.getSurgeFactor());
            receiptService.generateReceipt(rideId, fareBreakdown);

            // Free up driver
            if (ride.getDriver() != null) {
                Driver driver = ride.getDriver();
                driver.setStatus("online");
                driverRepository.save(driver);

                pricingService.updateDemandMetrics("increment", "driver");
            }

            socketService.sendRideStatusUpdated(ride);

            log.info("Trip ended: rideId={}", rideId);
            return ride;
        } catch (Exception e) {
            log.error("Error ending trip: rideId={}", rideId, e);
            throw new RuntimeException("Failed to end trip: " + e.getMessage());
        }
    }

    /**
     * Cancel a ride
     */
    @Transactional
    public Ride cancelRide(String rideId, String reason) {
        try {
            Ride ride = rideRepository.findById(rideId)
                    .orElseThrow(() -> new RuntimeException("Ride not found"));

            ride.setStatus("cancelled");
            ride = rideRepository.save(ride);

            // Free up driver if assigned
            if (ride.getDriver() != null) {
                Driver driver = ride.getDriver();
                driver.setStatus("online");
                driverRepository.save(driver);

                pricingService.updateDemandMetrics("increment", "driver");
            }

            // Update demand metrics
            pricingService.updateDemandMetrics("decrement", "ride");

            socketService.sendRideStatusUpdated(ride);

            log.info("Ride cancelled: rideId={}, reason={}", rideId, reason);
            return ride;
        } catch (Exception e) {
            log.error("Error cancelling ride: rideId={}", rideId, e);
            throw new RuntimeException("Failed to cancel ride: " + e.getMessage());
        }
    }
}
